addons/enemybar/actionTracking.lua

```diff
@@ -0,0 +1,299 @@
+	local is_pet = mob.is_npc and mob.charmed
+
+	local party = windower.ffxi.get_party()
+	for i=0, party.party1_count-1 do
+		local p = party['p'..i]
+		if p and p.mob then
+			if is_pet then
+				if p.mob.pet_index == mob.index then return true end
+			else
+				if p.mob.id == mob.id then return true end
+			end
+		end
+	end
+	for i=0, party.party2_count-1 do
+		local p = party['a1'..i]
+		if p and p.mob then
+			if is_pet then
+				if p.mob.pet_index == mob.index then return true end
+			else
+				if p.mob.id == mob.id then return true end
+			end
+		end
+	end
+	for i=0, party.party3_count-1 do
+		local p = party['a2'..i]
+		if p and p.mob then
+			if is_pet then
+				if p.mob.pet_index == mob.index then return true end
+			else
+				if p.mob.id == id then return true end
+			end
+		end
+	end
+
+    return false
+end
+
+function handle_action_packet(id, data)
+	if 0x028 == id then
+		local ai = windower.packets.parse_action(data)
+
+		track_enmity(ai)
+		track_actions(ai)
+		track_debuffs(ai)
+	elseif 0x029 then
+    	local message_id = data:unpack('H',0x19)
+    	if not message_id then return end
+    	message_id = message_id%32768
+		local param_1 = data:unpack('I',0x0D)
+    	local target_id = data:unpack('I',0x09)
+		if S{204,206}:contains(message_id) then
+			-- wears off message.
+			local target = windower.ffxi.get_mob_by_id(target_id)
+			if target and tracked_debuff[target.id]  then
+				tracked_debuff[target.id][param_1] = nil
+			end
+		end
+	end
+end
+
+function track_enmity(ai)
+	local actor = windower.ffxi.get_mob_by_id(ai.actor_id)
+	if not actor then return end
+
+	for i,t in ipairs(ai.targets) do
+		local target = windower.ffxi.get_mob_by_id(t.id)
+		if target then
+			local pc = nil
+			local mob = nil
+			if actor.is_npc and not actor.charmed and (not target.is_npc or target.charmed) then
+				pc = target
+				mob = actor
+			elseif (target.is_npc and not target.charmed) and (not actor.is_npc or target.charmed)  then
+				mob = target
+				pc = actor
+			end
+
+			if pc and mob and is_party_member_or_pet(pc) then
+				-- we have npc/pc interaction
+				-- if the actor is the npc, then we know there's enmity. Otherwise, if the target of the pc spell isn't tracked, it definitely hates the pc now.
+				if actor == mob or not tracked_enmity[mob.id] then
+					tracked_enmity[mob.id] = {pc=pc, mob=mob, time=os.time()}
+
+					-- if the actor is the npc, we don't care about the other targets. The first target is the one they targeted.
+					if actor == mob then return end
+				end
+			end
+		end
+	end
+end
+
+function track_actions(ai)
+	local actor = windower.ffxi.get_mob_by_id(ai.actor_id)
+	if not actor then return end
+
+	local target = windower.ffxi.get_mob_by_id(ai.targets[1].id)
+	if not target then
+		target = {name='Unknown (id:'..ai.targets[1].id, id=ai.targets[1].id}
+	end
+
+	-- if the category is not casting magic, jas, items or ws, don't bother.
+	if not S{8,4,7,11,3,6,9,5}:contains(ai.category) then return end
```

Comment:
This whole `S{1,2,3}:contains` thing has gotten a bit out of hand, and having started it I take full responsibility for it. This is also "slow", although I admit it is not enough to matter much in the end, it just bugs me personally (I would not mind it at all if you left it in). The problem with this is that it constructs the set each time the function is called. It constructs a new table with the values `{8, 4, 7, 11, 3, 6, 9, 5}` and then passes it to the `S` function, which copies all those values into a new table, which is then checked for with `:contains`.

Like I said, it's not a great optimization, but if you care for that kind of thing, you can define these fixed number sets outside of the function and just reuse them:
```lua
local ma_ja_item_ws = S{8, 4, 7, 11, 3, 6, 9, 5}
function track_actions(ai)
    --[[ ... ]]

    if not ma_ja_item_ws:contains(ai.category) then
        return
    end

    --[[ ... ]]
end
```